lsm/lsm.cpp

```diff
@@ -1,26 +1,93 @@
+#include <vector>
+#include <fstream>
+#include <limits>
+#include <bitset>
+#include "MurmurHash3.h"
 
 void BloomFilter::insert_key(int32_t k) {
-  auto key_hash = std::hash<int32_t>{}(k);
-  auto last_d = key_hash % 10;
-  bset[last_d] = 1;
+    uint32_t seed = 1;
+    uint64_t key_hash;
+    MurmurHash3_x86_32(&k, sizeof k, seed, &key_hash);
+    auto last_d = key_hash % 10;
+    bset[last_d] = 1;
 }
-
 char BloomFilter::contains_key(int32_t k) {
-  auto key_hash = std::hash<int32_t>{}(k);
-  auto last_d = key_hash % 10;
-  return bset[last_d];
+    uint32_t seed = 1;
+    uint64_t key_hash;
+    MurmurHash3_x86_32(&k, sizeof k, seed, &key_hash);
+    auto last_d = key_hash % 10;
+    return bset[last_d];
+}
+int Level::get_size() {
+  return files.size();
+}
+void Level::add_file(File f) {
+  files.push_back(f);
+}
+File Level::write_to_file(std::map<int32_t, maybe_value> m, std::string filename) {
+  std::ofstream ostrm(filename, std::ios::binary);
+  BloomFilter bloomFilter;
+  int min = std::numeric_limits<int>::max();
+  int max = std::numeric_limits<int>::lowest();
+  for (auto const&x : m) {
+    bloomFilter.insert_key(x.first);
+    ostrm.write(reinterpret_cast<const char*>(&x.first), sizeof x.first);
+    ostrm.write(reinterpret_cast<const char*>(&x.second.v), sizeof x.second.v);
+    ostrm.write(reinterpret_cast<const char*>(&x.second.is_deleted), sizeof x.second.is_deleted);
+    if (x.first > max) {max = x.first;}
+    if (x.first < min) {min = x.first;}
+  }
+  return File(filename, bloomFilter, min, max);
+}
+void Level::read_to_map(File f, std::map<int32_t, maybe_value> &m) { 
+  if (auto istrm = std::ifstream(f.filename, std::ios::binary)) {
+    int k;
+    while (istrm.read(reinterpret_cast<char *>(&k), sizeof k)) {
+      int v;
+      char d;
+      istrm.read(reinterpret_cast<char *>(&v), sizeof v);
+      istrm.read(reinterpret_cast<char *>(&d), sizeof d);
+      m.insert_or_assign(k, maybe_value(v,d));
+    }
+  } else {
+    std::cerr << "could not open " << f.filename << "\n";
+    exit(3);
+  }
+}
+File Level::merge(File older_f, File newer_f, std::string merged_filename) {
+  std::map<int32_t, maybe_value> m;
+  read_to_map(older_f, m);
+  read_to_map(newer_f, m);
+  return write_to_file(m, merged_filename);
 }
 
+void LSM::add_to_level(File f, int l_index) {
+  if ((size_t)l_index >= levels.size()) {
+    Level l = Level();
+    l.add_file(f);
+    levels.push_back(l);
+  } else if (levels[l_index].get_size() + 1 >= levels[l_index].max_size) {
+    File merged_f = levels[l_index].merge(levels[l_index].files.back(), f, "foo" );
+    add_to_level(merged_f, l_index + 1);
+    levels[l_index].files.pop_back();
+  } else {
+    levels[l_index].add_file(f);
+  }
+}
 void LSM::put(int32_t k, int32_t v, char deleted) {
   if (m.size() >= 4) {
     std::string filename = directory / std::to_string(files.size());
-    files.push_back(write_to_file(filename));
+    File f = write_to_file(filename);
+    files.push_back(f);
```

Comment:
Do you still need LSM.files? Each file is tracked by the level now. You can have the Level class manage file names. If each level knows its index in `levels` then it could return new file names as: `directory / std::to_string(my_level_index) / std::to_string(files.size())`

Reply:
You can also use this in place of `"foo"` above when you need a merged filename.

It's OK to re-use filenames within a given level. As soon as you call `levels[l_index].files.pop_back` that file is meaningless to you.